File: 02_UML/Object_Association/Composition/Main.cpp


/*

Composition is a strong whole-part relationship in which the part's lifetime 
is controlled by the whole; when the whole is destroyed, its parts are also destroyed.

*/

#include <iostream>
using namespace std;

class OrderItem {
public:
    OrderItem(string product) {
        cout << product << " created\n";
    }

    ~OrderItem() {
        cout << "OrderItem destroyed\n";
    }
};

class Order {
private:
    OrderItem item1;
    OrderItem item2;

public:
    Order()
        : item1("Laptop"),
          item2("Mouse") {
    }

    ~Order() {
        cout << "Order destroyed\n";
    }
};

int main() {

    {
        Order order;
    }

    // order is destroyed here
    // item1 and item2 are also destroyed
}

File: 02_UML/Object_Association/Composition/Main.java
import java.util.ArrayList;
import java.util.List;

class OrderItem {

    private String productName;

    OrderItem(String productName) {
        this.productName = productName;
    }
}

class Order {

    private List<OrderItem> items = new ArrayList<>();

    public Order() {
        items.add(new OrderItem("Laptop"));
        items.add(new OrderItem("Mouse"));
    }
}

public class Main {

    public static void main(String[] args) {

        Order order = new Order();

    } // order becomes unreachable here,
      // and its OrderItems are also owned by it.
}